### ISSUE TITLE
Record<string, string> no longer compatible with Wrapper<T extends string> = Record<T, string>

### ISSUE THREAD

**Issue by ejose19** [microsoft/TypeScript#42931]
# Bug Report

### 🔎 Search Terms

Record generic
String Record

### 🕗 Version & Regression Information

- This changed between versions 4.1.5 and 4.2.2

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.2.0-beta#code/PTAEFkE9QYwewCYFNQHM5IM6gBZIE5IBQALpAA4oDq+AhuZfgDwAqoSAHiUgHYLaYS+AJY9UAPlABeUACUk8fAlYAaUIJFjxAblIUUAaySRsMgOQAzOHDO6i8HoNAAjWpiQAuOQrhKmG0VQ1AK1pUABvUCs4LzMzUABfXQcnOGcAKwBGL3lFZSMTYKFAyRlXd21QEFAeOHZ8fF9sN3YOShhuBHs4RxJQNPSAJi8aekYmAsxSlzckSuqCRvxsAHdhEhxQAAMdgAVGxjJQS2t44WwAW3PMQNBRUDJKY9zfZRCg9WKteOcAVz7CABHX7CQgIO48B76Y6jBgEJgAImiCPEZgAdDstkA)

### 💻 Code

```ts
type Wrapper<T extends string> = Record<T, string>;
type keys = 'foo';

const base: Record<string, string> = { foo: '' };
const obj1: Record<keys, string> = base; // no errors as expected
const obj2: Wrapper<keys> = base; // errors with ```Property 'foo' is missing in type 'Record<string, string>' but required in type 'Wrapper<"foo">'.```
```

### 🙁 Actual behavior

Typescript errors with ```Property 'foo' is missing in type 'Record<string, string>' but required in type 'Wrapper<"foo">'.```

### 🙂 Expected behavior

No errors should be given (as in previous versions) since this is just a wrapper type.

### Note:

Even this "workaround" gives the expected result:

```ts
interface Wrapper<T extends string> {
  workaround: Record<T, string>
}
type keys = 'foo';

const base: Record<string, string> = { foo: '' };
const obj: Wrapper<keys>['workaround'] = base; // no errors
```

so the issue seems to be affecting "type" only

**Comment by IllusionMH** [comment]
IMHO ideally it should be other way around with both `obj1` and `obj2` being errors.
[Example](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210224#code/MYewdgzgLgBAhmAnlAFgSzAcwFwwEoCmoATgCYA80xGmANDFTQHwwC8MA3jMLgOTC96pPqV4wAvgG4AUNNCRYAdzSoAggCFchEhV5wxAHxi8ARoIZRqWFuwTJ0WSTAD0zmGBAwCxYiGIxsXgAKWXkIEAAbAgA6CJBMIOU1dWjAQRBoqBAAGRBgOCiAVQAHIu8AYTgIAiCAShqnV24-YiIoCMQvHz8IegIANwIwGDQAM24vAA80aAhQ8HCo2PjElRQNaLgM7Nz8gmLS4gqq2vqXNw9O32J6EwBXWGBiSpQCWaA)
```ts
const anything: Record<string, string> = { c: 'c', d: 'd' };

const withAB: Record<'a' | 'b', string> = anything; // no error :'(

console.log(withAB.с.toLocaleUpperCase()); // correctly errors, even if c exists

console.log(withAB.a.toLocaleUpperCase()); // no error, but crashes
```

**Comment by ejose19** [comment]
You would need to break the ```Record<string, string>``` completely, as even without the reported issue, you can do this:

```ts
const anything: Record<string, string> = { c: 'c', d: 'd' };
console.log(anything.a.toLocaleUpperCase());
```

**Comment by IllusionMH** [comment]
Sorry, I forgot that `noUncheckedIndexedAccess` is not enabled by default and not enabled in playground as well.

`anything.a` case already solved by `"noUncheckedIndexedAccess": true` but assignment to `Record<'a' | 'b', string>` still allowed.

And I agree that, my comment is not relevant to case with `"noUncheckedIndexedAccess": false`.

**Comment by ejose19** [comment]
@IllusionMH Correct, even with ```"noUncheckedIndexedAccess": true``` it allows assigning ```const anything2: Record<'a' | 'b', string> = anything``` which shouldn't be allowed (as after that ```anything2.a``` is assumed to be non undefined)

**Comment by RyanCavanaugh** [comment]
I could make convincing arguments in both directions but it's very surprising to have these diverge. I think we took a wrong turn at a variance measurement on `Wrapper`.

**Comment by orta** [comment]
I ran this through with Wes, we think the change which triggered this is the better tracking of types throughout the system. What happened is that this change went from being a variance check in the old system (Record -> Record) to a structural check (Wrapper -> Record) where it failed. Here's an example of a completely [unwrapped version of what's happening](https://www.typescriptlang.org/play?ts=4.1.5#code/C4TwDgpgBAShDGB7ATgEwEwB4DSUIA9gIA7VAZygGsIREAzKAQ2JABooAVAPigF4oA3gCgooqAG0AClACWxKNgC6ALk4BuIQF8NoSFRoV+AcjqJERjUKTEywKACNGZCKoD0AKjhI0WW8jkA5ux+gVzuroISMqgEqiHEASpQ8QGafJGmiKpGRlDaVog2doj2AFYAjG6eCCiomNQgZMHA-glhEQKZcS2BafyOzmpAA)

```ts
type Record2<K extends keyof any, T> = {
    [P in K]: T;
};
type keys = 'foo';

const base: /*Record2<string, string>*/ {[idex: string]: string} = { foo: '' };
const obj1: /*Record<keys, string>*/ {foo: string} = base;
      ^^^^
// Property 'foo' is missing in type '{ [idex: string]: string; }' but required in type '{ foo: string; }'.(2741)
```

We don't think this new behavior is wrong, but it's just a well hidden surprise for why TS errors

**Comment by orta** [comment]
I think there's not really any actions on our side here, the fails represents not applying the structural analysis prior to 4/2 and we're doing a better  analysis now
